File: typesense-gin-full-text-search/search/collections.go
package search

import (
	"context"
	"fmt"
	"log"

	"github.com/typesense/code-samples/typesense-gin-full-text-search/config"
	"github.com/typesense/typesense-go/v4/typesense/api"
	"github.com/typesense/typesense-go/v4/typesense/api/pointer"
)

// InitializeCollections ensures all required collections exist
// This is idempotent - safe to call multiple times
func InitializeCollections(ctx context.Context) error {
	log.Println("Initializing Typesense collections...")

	booksSchema := &api.CollectionSchema{
		Name: config.BookCollection,
		Fields: []api.Field{
			{Name: "title", Type: "string", Facet: pointer.False()},
			{Name: "authors", Type: "string[]", Facet: pointer.True()},
			{Name: "publication_year", Type: "int32", Facet: pointer.True()},
			{Name: "average_rating", Type: "float", Facet: pointer.True()},
			{Name: "image_url", Type: "string", Facet: pointer.False()},
			{Name: "ratings_count", Type: "int32", Facet: pointer.True()},
		},
		DefaultSortingField: pointer.String("ratings_count"),
	}

	_, err := Client.Collection(config.BookCollection).Retrieve(ctx)
	if err != nil {
		log.Printf("Collection '%s' not found, creating...", config.BookCollection)
		_, err = Client.Collections().Create(ctx, booksSchema)
		if err != nil {
			return fmt.Errorf("failed to create collection '%s': %w", config.BookCollection, err)
		}
		log.Printf("Collection '%s' created successfully", config.BookCollection)
	} else {
		log.Printf("Collection '%s' already exists, skipping creation", config.BookCollection)
	}

	return nil
}

// CollectionDocumentCount returns the number of documents in the books collection.
// Returns 0 on any error (treated as empty).
func CollectionDocumentCount(ctx context.Context) int64 {
	coll, err := Client.Collection(config.BookCollection).Retrieve(ctx)
	if err != nil || coll.NumDocuments == nil {
		return 0
	}
	return *coll.NumDocuments
}

File: typesense-node-drizzle-full-text-search/src/search/collections.ts
import { typesenseClient } from './client';
import { env } from '../config/env';
import type { CollectionCreateSchema } from 'typesense/lib/Typesense/Collections';

export const BOOKS_COLLECTION_NAME = env.TYPESENSE_COLLECTION;

export async function initializeTypesense() {
  const schema: CollectionCreateSchema = {
    name: BOOKS_COLLECTION_NAME,
    fields: [
      { name: 'title', type: 'string', facet: false },
      { name: 'authors', type: 'string[]', facet: true },
      { name: 'publication_year', type: 'int32', facet: true },
      { name: 'average_rating', type: 'float', facet: true },
      { name: 'image_url', type: 'string', facet: false },
      { name: 'ratings_count', type: 'int32', facet: true },
    ],
    default_sorting_field: 'ratings_count',
  };

  try {
    await typesenseClient.collections(BOOKS_COLLECTION_NAME).retrieve();
    console.log(`Collection '${BOOKS_COLLECTION_NAME}' already exists.`);
  } catch (error: any) {
    if (error.httpStatus === 404) {
      console.log(`Collection '${BOOKS_COLLECTION_NAME}' not found. Creating...`);
      await typesenseClient.collections().create(schema);
      console.log(`Collection '${BOOKS_COLLECTION_NAME}' created successfully.`);
    } else {
      throw error;
    }
  }
}
